sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_crypto.c

```diff
@@ -55,25 +97,104 @@ void gnrc_lorawan_calculate_join_mic(const uint8_t *buf, size_t len,
     memcpy(out, digest, sizeof(le_uint32_t));
 }
 
-void gnrc_lorawan_calculate_mic(const le_uint32_t *dev_addr, uint32_t fcnt,
-                                uint8_t dir, iolist_t *frame,
-                                const uint8_t *nwkskey, le_uint32_t *out)
+#if IS_USED(CONFIG_LORAWAN_1_1)
```

Comment:
Same here. You can merge these two functions and use a common signature. If you have different parameters, consider a structure.
You can also kind of merge the logic here: `gnrc_lorawan_calculate_mic_uplink_11x` is (almost) a subset of `gnrc_lorawan_calculate_mic_uplink_10x`.

Also, you can use a regular `if` with `IS_ACTIVE`or `IS_USED` in some parts. The compiler will remove branches that are never reached. Of course, you cannot do that with struct members that don't exist on compile time (does who appear only with LoRaWAN 1.1).

For instance, you could write this functions as:
```c
void gnrc_lorawan_calculate_mic_uplink(const le_uint32_t *dev_addr, uint32_t fcnt,
                                           ... lorawan_mic_t *ctx, ...
                                           le_uint32_t *out)
{
    orawan_block0_t block0 = { 0 };

    block0.fb = MIC_B0_START;
    block0.dir = 0x00;
        
    memcpy(&block0.dev_addr, dev_addr, sizeof(le_uint32_t));

    block0.fcnt = byteorder_htoll(fcnt);
    block0.len = iolist_size(frame);

    /* cmacF = aes128_cmac(FNwkSIntKey, B0 | msg) */
    cmac_init(&CmacContext, fnwksintkey, LORAMAC_FNWKSINTKEY_LEN);
    cmac_update(&CmacContext, &block0, sizeof(block0));
    for (iolist_t *io = frame; io != NULL; io = io->iol_next) {
        cmac_update(&CmacContext, io->iol_base, io->iol_len);
    }
    cmac_final(&CmacContext, digest);
  
    if (IS_ACTIVE(GNRC_LORAWAN_1_1)) {
        /* Do 1.1 specific stuff an get specific keys from `ctx`*/
    }
}
```

Reply:
When exactly can I use `if` with this macro ? I tried to e.g. use [here](https://github.com/Ollrogge/RIOT/blob/74caa2caa3a125d90b67124481d5ec46c57f5e16/sys/net/gnrc/netif/lorawan/gnrc_netif_lorawan.c#L277) and it didn't work.
I find it kinda weird to depend on compiler optimizations to make my code compile ...

Reply:
it probably failed because `_appkey` exists only when you set LoRaWAN 1.1. Since the "topology" of the file is the same, the variables should always exist.

You can always use `if` with this macro, since always resolves to 0 or 1.

For example:
```c
/* foo */
if (IS_ACTIVE(MY_MODULE)) {
    /* bar */
}
```
In this example the snippets "foo" and "bar" are compiled like regular C code. But in case "MY_MODULE" is not used, the compiler will optimize out "bar". The compiler will still find syntax errors in "bar", since the compiler parses the snippet anyway.

Reply:
> I find it kinda weird to depend on compiler optimizations to make my code compile ...

What do you mean with that? You don't depend on compiler optimizations to compile your code (if you turn off compiler optimizations it will still compile without issues). Could you elaborate more?

Reply:
> it probably failed because `_appkey` exists only when you set LoRaWAN 1.1. Since the "topology" of the file is the same, the variables should always exist.
> 
> You can always use `if` with this macro, since always resolves to 0 or 1.
> 
> For example:
> 
> ```c
> /* foo */
> if (IS_ACTIVE(MY_MODULE)) {
>     /* bar */
> }
> ```
> 
> In this example the snippets "foo" and "bar" are compiled like regular C code. But in case "MY_MODULE" is not used, the compiler will optimize out "bar". The compiler will still find syntax errors in "bar", since the compiler parses the snippet anyway.

Having `_appkey` exist all the time fixed the issue. Thank you ! 
I misunderstood how using `if (IS_USED(...)` works. Your example and explanation cleared things up !